### ISSUE TITLE
Live Event Detection for People’s Safety Using NLP and Deep Learning

### ISSUE THREAD

**Issue by surya190100** [surya190100/Inventory-management-system#1]
Ensuring public safety at live events is a critical challenge, especially with the increasing
frequency of large gatherings, such as concerts, sports events, and festivals. Live event
detection involves identifying safety risks in real time, enabling authorities to intervene
promptly and mitigate potential threats. Among various data types, audio signals offer a rich
source of information, as they can capture crowd reactions, abnormal sounds, or potential
hazards such as explosions or disturbances. Advanced machine learning methods provide a
powerful way to analyze these signals and detect anomalies effectively.
The proposed work focuses on designing a framework for real-time live event detection using
audio data. By incorporating advanced feature extraction techniques and the robust
classification capabilities of LightGBM, this system aims to achieve high accuracy in
identifying abnormal events. This method not only addresses the limitations of existing
approaches but also offers scalability for deployment across various event types and scales

**Comment by surya190100** [comment]
[Live Event Sound.docx](https://github.com/user-attachments/files/20907307/Live.Event.Sound.docx)
[MAJOR\[Live_Event_Detection_for_People’s_Safety_Using_NLP_nad_DEEP_LEARNING_(WITHOUT CERTIFICATE)\].pdf](https://github.com/user-attachments/files/20907305/MAJOR.Live_Event_Detection_for_People.s_Safety_Using_NLP_nad_DEEP_LEARNING_.WITHOUT.CERTIFICATE.pdf)
[requirement.txt](https://github.com/user-attachments/files/20907306/requirement.txt)
![Image](https://github.com/user-attachments/assets/bb458a96-cefc-43f6-aad7-156095062f22)

**Comment by surya190100** [comment]
`# A Comprehensive Dataset for Live event sound Classification and Analysis
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.utils import resample
import librosa
import os
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.decomposition import PCA
import IPython
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import precision_score,accuracy_score,f1_score,recall_score
from sklearn.metrics import classification_report, confusion_matrix
import warnings
warnings.filterwarnings('ignore')
from sklearn.neural_network import MLPClassifier
from lightgbm import LGBMClassifier
import joblib
from imblearn.over_sampling import SMOTE
path = r"sorted_audio_data"
model_folder = "model"
categories = [d for d in os.listdir(path) if os.path.isdir(os.path.join(path, d))]
categories
def remove_noise(audio, sr):
 noise_profile = audio[:int(0.5 sr)]
 noise_reduced_audio = librosa.effects.remix(audio, intervals=librosa.effects.split(audio,
top_db=20))
 return noise_reduced_audio
def load_audio_from_directory_or_cache(path, categories, sr, n_mfcc, model_folder):
 X_file = os.path.join(model_folder, "X.npy")
 Y_file = os.path.join(model_folder, "Y.npy")
 if os.path.exists(X_file) and os.path.exists(Y_file):
 print(f"Loading cached data from {model_folder}")
 X = np.load(X_file)
 Y = np.load(Y_file)
 else:
 print(f"Path does not exist: {path}" if not os.path.exists(path) else "Processing
directory")
 X = []
 Y = []
 if not os.path.exists(path):
 print(f"Path does not exist: {path}")
 return np.array(X), np.array(Y)
 y, sr = librosa.load(file_path, sr=sr)
 y = remove_noise(y, sr)
 mfccs = librosa.feature.mfcc(y=y, sr=sr, n_mfcc=30)
 mfccs_mean = np.mean(mfccs.T, axis=0) # Taking the mean of the MFCCs
 X.append(mfccs_mean)
 if name in categories:
 Y.append(categories.index(name))
 else:
 print(f"Category {name} not in categories list.")
 except Exception as e:
 print(f"Skipping {file_path}, error reading file: {e}")
 X = np.array(X)
 Y = np.array(Y)
 os.makedirs(model_folder, exist_ok=True) # Create the directory if it doesn't exist
# Load audio
X, Y = load_audio_from_directory_or_cache(path, categories, 22050, 13, 'model') [cite: 617, 618]
# Debug print statements to check the shape of X and Y [cite: 619]
print(f"Shape of X before any reshaping: {X.shape}") [cite: 619]
print(f"Shape of Y: {Y.shape}") [cite: 619]
# Check the shape of X and ensure it is 2D [cite: 619]
if len(X.shape) != 2: [cite: 619]
 raise ValueError(f"X should be a 2D array but got shape {X.shape}") [cite: 619]
# Normalize features [cite: 619]
scaler = StandardScaler() [cite: 619]
X = scaler.fit_transform(X) [cite: 619]
# Save arrays [cite: 619]
print(f"Processed and saved {len(X)} audio files.") [cite: 619]
X_list = X.tolist() [cite: 619]
Y_list = Y.tolist() [cite: 619]
# Create DataFrame [cite: 619]
df = pd.DataFrame({'X': X_list, 'Y': Y_list}) [cite: 619]
# Plot the counts of each category [cite: 619]
category_counts = {category: len(os.listdir(os.path.join(path, category))) for category in
categories} [cite: 620]
df_counts = pd.DataFrame(list(category_counts.items()), columns=['Category', 'Count']) [cite: 620]
print(algorithm+' Accuracy : '+str(a)) [cite: 620]
print(algorithm+' Precision : '+str(p)) [cite: 620]
print(algorithm+' Recall : '+str(r)) [cite: 620]
print(algorithm+' FSCORE : '+str(f)) [cite: 620]
report=classification_report(predict, testY,target_names=labels) [cite: 620]
print('\n',algorithm+" classification report\n",report) [cite: 620]
conf_matrix = confusion_matrix(testY, predict)
plt.figure(figsize =(5, 5))
ax = sns.heatmap(conf_matrix, xticklabels = labels, yticklabels = labels, annot = True,
cmap="Blues" ,fmt ="g"); [cite: 621]
.xlabel('Predicted class') [cite: 621]
plt.show() [cite: 621]
lgbm = LGBMClassifier() [cite: 621]
lgbm.fit(X_train, y_train) [cite: 621]
joblib.dump(lgbm,path) [cite: 621]
y_pred = lgbm.predict(X_test) [cite: 621]
performance_metrics('LGBMClassifier', y_pred, y_test) [cite: 621]
# prediction on new test data [cite: 621]
# Prediction on new test data [cite: 621]
def preprocess_audio(audio_path, scaler): [cite: 621]
 audio, sr = librosa.load(audio_path, sr=None) [cite: 621]
 audio = remove_noise(audio, sr) [cite: 621]`
